### PR TITLE
Support Spark 3.4

### LINKED ISSUE #1696: [Feature Request] Support Spark 3.4

### PR DESCRIPTION
**Pull request by allisonport-db**
## Description

Makes changes to support Spark 3.4. These include compile necessary changes, and test _and_ code changes due to changes in Spark behavior.

Some of the bigger changes include
- A lot of changes regarding error classes. These include...
  - Spark 3.4 changed `class ErrorInfo` to private. This means the current approach in `DeltaThrowableHelper` can no longer work. We now use `ErrorClassJsonReader` (these are the changes to `DeltaThrowableHelper` and `DeltaThrowableSuite`
  - Many error functions switched the first argument from `message: String` to `errorClass: String` which **does not** cause a compile error, but instead causes a "SparkException-error not found" when called. Some things affected include `ParseException(...)`, `a.failAnalysis(..)`.
  - Supports error subclasses
- Spark 3.4 supports insert-into-by-name and no longer reorders such queries to be insert-into-by-ordinal. See https://github.com/apache/spark/pull/39334. In `DeltaAnalysis.scala` we need to perform schema validation checks and schema evolution for such queries; right now we only match when `!isByName`
- SPARK-27561 added support for lateral column alias. This broke our generation expression validation checks for generated columns. We now separately check for generated columns that reference other generated columns in `GeneratedColumn.scala`
- `DelegatingCatalogExtension` deprecates `createTable(..., schema: StructType, ...)` in favor of `createTable(..., columns: Array[Column], ...)`
- `_metadata.file_path` is not always encoded. We update `DeleteWithDeletionVectorsHelper.scala` to accomodate for this.
- Support for SQL `REPLACE WHERE`. [TESTS IN FOLLOW-UP PR]
-  Misc test changes due to minor changes in Spark behavior or error messages

Resolves #1696

## How was this patch tested?

Existing tests should suffice since there are no major Delta behavior changes _besides_ support for `REPLACE WHERE` for which we have added tests.

## Does this PR introduce _any_ user-facing changes?

Yes. Spark 3.4 will be supported. `REPLACE WHERE` is supported in SQL.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala [hunk 1]
```diff
@@ -750,6 +769,58 @@ class DeltaAnalysis(session: SparkSession)
...
+
+  /**
+   * With Delta, we ACCEPT_ANY_SCHEMA, meaning that Spark doesn't automatically adjust the schema
+   * of INSERT INTO. Here we check if we need to perform any schema adjustment for INSERT INTO by
+   * name queries. We also check that any columns not in the list of user-specified columns must
+   * have a default expression.
+   */
+  private def needsSchemaAdjustmentByName(query: LogicalPlan, targetAttrs: Seq[Attribute],
+      deltaTable: DeltaTableV2): Boolean = {
+    // TODO: update this to allow columns with default expressions to not be
```
Comment (allisonport-db) [inline core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:807]: This will come in a follow-up PR with tests

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaSharedExceptions.scala [hunk 2]
```diff
@@ -16,13 +16,23 @@
 
 package org.apache.spark.sql.delta
 
+import org.antlr.v4.runtime.ParserRuleContext
+
 import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.parser.{ParseException, ParserUtils}
+import org.apache.spark.sql.catalyst.trees.Origin
 
 class DeltaAnalysisException(
```
Comment (allisonport-db) [inline core/src/main/scala/org/apache/spark/sql/delta/DeltaSharedExceptions.scala:25]: We add these parameters so we don't lose information in our exceptions thrown in `deltaMerge.scala` that used to use `a.failAnalysis(...)`

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaSharedExceptions.scala [hunk 3]
```diff
@@ -48,3 +58,17 @@ class DeltaUnsupportedOperationException(
     override def getErrorClass: String = errorClass
     def getMessageParametersArray: Array[String] = messageParameters
 }
+
+// todo: we had to add this since in Spark 3.4 ParseException(message, ...) was replaced by
+//   ParseException(errorClass, ...)
+//   Instead of passing just a message here, we could enforce creating an errorClass for each
+//   invocation and make this DeltaParseException(errorClass, ctx)
+class DeltaParseException(
```
Comment (allisonport-db) [inline core/src/main/scala/org/apache/spark/sql/delta/DeltaSharedExceptions.scala:66]: In Spark 3.3 there is a constructor `ParseException(message, ctx)` but in 3.4 it is `ParseException(errorClass, ctx)` (and that error class would need to be a _Spark error class_).

Here we enable creating a `ParseException` from a message and context.

File: core/src/main/scala/org/apache/spark/sql/delta/commands/WriteIntoDelta.scala [hunk 4]
```diff
@@ -198,8 +198,7 @@ case class WriteIntoDelta(
       }
     }
     val rearrangeOnly = options.rearrangeOnly
-    // TODO: use `SQLConf.READ_SIDE_CHAR_PADDING` after Spark 3.4 is released.
-    val charPadding = sparkSession.conf.get("spark.sql.readSideCharPadding", "false") == "true"
+    val charPadding = sparkSession.conf.get(SQLConf.READ_SIDE_CHAR_PADDING.key, "false") == "true"
```
Comment (allisonport-db) [inline core/src/main/scala/org/apache/spark/sql/delta/commands/WriteIntoDelta.scala:201]: The default in Spark is "true" so we cannot use `sparkSession.conf.get(SQLConf.READ_SIDE_CHAR_PADDING)`

File: core/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/deltaMerge.scala [hunk 5]
```diff
@@ -424,7 +424,12 @@ object DeltaMergeInto {
         // Note: This will throw error only on unresolved attribute issues,
         // not other resolution errors like mismatched data types.
         val cols = "columns " + plan.children.flatMap(_.output).map(_.sql).mkString(", ")
-        a.failAnalysis(msg = s"cannot resolve ${a.sql} in $mergeClauseType given $cols")
+        // todo: added a new Delta error for this to avoid rewriting tests, but existing
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/deltaMerge.scala:427]: Will there be a follow up PR for this? If so, please clarify what need to be done. This is a little vague for a TODO
```suggestion
        // TODO: added a new Delta error for this to avoid rewriting tests, but existing
```

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala [hunk 6]
```diff
@@ -85,6 +86,18 @@ class DeltaAnalysis(session: SparkSession)
       }
 
 
+    // INSERT INTO by name
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:89]: is this new functionality? if not, how did the the corresponding command work before with this case statement? What changed with Spark 3.4?

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala [hunk 7]
```diff
@@ -750,6 +769,58 @@ class DeltaAnalysis(session: SparkSession)
       !SchemaUtils.isReadCompatible(schema.asNullable, existingSchemaOutput.toStructType)
   }
 
+  private def insertIntoByNameMissingColumn(
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:772]: Please add method docs.

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala (hunk 6 above)
Comment (tdas) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:89]: Aah i get it now.. this is the inverse of the condition above ... for `a.isByName`. Before this change, did this throw an error? If so, then this is new support for new SQL commands.. right?

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaSharedExceptions.scala [hunk 8]
```diff
@@ -48,3 +58,17 @@ class DeltaUnsupportedOperationException(
     override def getErrorClass: String = errorClass
     def getMessageParametersArray: Array[String] = messageParameters
 }
+
+// todo: we had to add this since in Spark 3.4 ParseException(message, ...) was replaced by
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/DeltaSharedExceptions.scala:62]: nit: todo's are generally capitalized.

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala (hunk 1 above)
Comment (tdas) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:807]: What does this mean? Are there any current tests covering this function... or is the entire thing TODO?

Comment (tdas) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:807]: The last thing we want is some half-incorrectly-implemented functionality.

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala [hunk 9]
```diff
@@ -759,15 +830,16 @@ class DeltaAnalysis(session: SparkSession)
           Cast(input, dt, Option(timeZone), ansiEnabled = false)
       case SQLConf.StoreAssignmentPolicy.ANSI =>
         (input: Expression, dt: DataType, name: String) => {
-          AnsiCast(input, dt, Option(timeZone))
+          val cast = Cast(input, dt, Option(timeZone), ansiEnabled = true)
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:833]: What is this change for?

File: core/src/main/scala/org/apache/spark/sql/delta/GeneratedColumn.scala [hunk 10]
```diff
@@ -161,6 +162,37 @@ object GeneratedColumn extends DeltaLogging with AnalysisHelper {
...
 
+  /**
+   * SPARK-27561 added support for lateral column alias. This means generation expressions that
+   * reference other generated columns no longer fail analysis in `validateGeneratedColumns`.
+   *
+   * This method checks for and throws an error if:
+   * - A generated column references itself
+   * - A generated column references another generated column
+   */
+  def validateColumnReferences(
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/GeneratedColumn.scala:173]: is there test coverage for this?

File: core/src/test/scala/org/apache/spark/sql/delta/HiveDeltaDDLSuite.scala [hunk 11]
```diff
@@ -26,11 +26,6 @@ abstract class HiveDeltaDDLSuiteBase
   extends DeltaDDLTestBase {
   import testImplicits._
 
-  override protected def verifyDescribeTable(tblName: String): Unit = {
```
Comment (tdas) [inline core/src/test/scala/org/apache/spark/sql/delta/HiveDeltaDDLSuite.scala:29]: why is this removed? not used any more?

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala (hunk 9 above)
Comment (allisonport-db) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:833]: Compile error; `AnsiCast` removed in 3.4 and consolidated into one `Cast` class

File: core/src/main/scala/org/apache/spark/sql/delta/GeneratedColumn.scala (hunk 10 above)
Comment (allisonport-db) [reply core/src/main/scala/org/apache/spark/sql/delta/GeneratedColumn.scala:173]: Yes in `GeneratedColumnSuite`, tests fail without this change.

File: build.sbt [hunk 12]
```diff
@@ -198,23 +198,24 @@ lazy val storageS3DynamoDB = (project in file("storage-s3-dynamodb"))
...
-        val (expMaj, expMin, _) = getMajorMinorPatch(sparkVersion)
-        ("org.apache.iceberg" % s"iceberg-spark-runtime-$expMaj.$expMin" % "1.1.0" % "provided")
-          .cross(CrossVersion.binary)
-      },
-      // Fix Iceberg's legacy java.lang.NoClassDefFoundError: scala/jdk/CollectionConverters$ error
-      // due to legacy scala.
-      "org.scala-lang.modules" %% "scala-collection-compat" % "2.1.1"
-    )
-  )
+// Requires iceberg release on 3.4
```
[the remark this answers is not in the document]
Comment (allisonport-db) [reply build.sbt:201]: Yeah sorry; I originally just commented this out to avoid breaking the build

File: core/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/deltaMerge.scala (hunk 5 above)
Comment (allisonport-db) [reply core/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/deltaMerge.scala:427]: Ah just a note to myself / reviewers on the other option. I think I prefer using the new Delta error I will remove this comment

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala (hunk 6 above)
Comment (allisonport-db) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:89]: Spark 3.3:
- insert into by name queries
    - if all columns specified --> rearrange order and convert to an insert by ordinal query
    - not all column specified --> exception (missing columns)
   
In Spark 3.4 insert into by name queries for DSV2 are not converted to by ordinal https://github.com/apache/spark/pull/39334

So before insert by name queries would be converted to insert by ordinal and would match the above `AppendDelta` pattern.

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala (hunk 1 above)
Comment (allisonport-db) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:807]: We add the code to support the new functionality (not specifying generated columns in insert into by name statements), but we don't allow it in this PR. Basically in Spark 3.3, insert into by name only worked when all the columns were specified and otherwise would throw an error. We maintain that behavior here (lines 809-812) so no change in behavior (so existing tests should suffice.)

Once we remove the check on 809-812 we will have the full support for insert into with generated columns. I'll remove it and add the tests in a follow-up PR.

Comment (allisonport-db) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:807]: Since we have to now do schema validation / evolution for by-name queries it's easier to add all this code now, but artificially block new functionality so the tests can come in a follow-up PR and lessen the amount of code in this PR.

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala (hunk 6 above)
Comment (allisonport-db) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:89]: This is mentioned in the PR description albeit a little less thoroughly :)

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaSharedExceptions.scala (hunk 8 above)
Comment (allisonport-db) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaSharedExceptions.scala:62]: Sorry also didn't plan to merge this todo. Mostly a note to myself or reviewers. I think it's easiest not to add an error class for every invocation in `DeltaSqlParser` but I'm not super clear on how much we want to stick to enforcing using the Spark error framework. I'm happy to do it if we want to stick to the framework

File: core/src/test/scala/org/apache/spark/sql/delta/HiveDeltaDDLSuite.scala (hunk 11 above)
Comment (allisonport-db) [reply core/src/test/scala/org/apache/spark/sql/delta/HiveDeltaDDLSuite.scala:29]: Sorry should've added a note here. This is the same in all child suites so I moved the definition to `DeltaDDLTestBase` [here](https://github.com/delta-io/delta/pull/1720/files#diff-90717098e565cbbb91601b34e60bc6aca86c6f9ee6e6fd47b5dac75c735a0841L96) and removed them from the subclasses.

File: python/delta/tests/test_pip_utils.py [hunk 13]
```diff
@@ -73,10 +73,12 @@ def tearDown(self) -> None:
         shutil.rmtree(self.tempPath)
 
     def test_maven_jar_loaded(self) -> None:
-        packages: List[str] = self.spark.conf.get("spark.jars.packages").split(",")
-
+        packagesConf: Optional[str] = self.spark.conf.get("spark.jars.packages")
```
[the remark this answers is not in the document]
Comment (allisonport-db) [reply python/delta/tests/test_pip_utils.py:76]: 3.4 changed the return type from `conf.get` from `str` --> `Optional[str]`
mypi checks fail without this specific assert statement

File: python/delta/tests/test_pip_utils.py [hunk 14]
```diff
@@ -73,10 +73,12 @@ def tearDown(self) -> None:
...
+        assert packagesConf is not None  # mypi needs this to assign type str
```
Comment (allisonport-db) [inline python/delta/tests/test_pip_utils.py:77]:
```suggestion
        assert packagesConf is not None  # mypi needs this to assign type str from Optional[str]
```
does this help? since the types are with variable definition i don't know if docs mentioning the return type will help

Comment (allisonport-db) [reply python/delta/tests/test_pip_utils.py:77]: Maybe there is a better way to do this I had to google this

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala (hunk 7 above)
Comment (allisonport-db) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:778]: Added

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala (hunk 6 above)
Comment (tdas) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:89]: understood. thankyou for explaining in detail

File: core/src/test/scala/org/apache/spark/sql/delta/HiveDeltaDDLSuite.scala (hunk 11 above)
Comment (tdas) [reply core/src/test/scala/org/apache/spark/sql/delta/HiveDeltaDDLSuite.scala:29]: nice!

File: core/src/main/scala/org/apache/spark/sql/delta/commands/DeleteWithDeletionVectorsHelper.scala [hunk 15]
```diff
@@ -293,8 +294,11 @@ object DeletionVectorBitmapGenerator {
       candidateFiles: Seq[AddFile],
       condition: Expression)
     : Seq[DeletionVectorResult] = {
+    val uriEncode = DeltaUDF.stringFromString(path => {
+      new Path(path).toUri.toString
+    })
     val matchedRowsDf = targetDf
-      .withColumn(FILE_NAME_COL, col(s"${METADATA_NAME}.${FILE_PATH}"))
+      .withColumn(FILE_NAME_COL, uriEncode(col(s"${METADATA_NAME}.${FILE_PATH}")))
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/commands/DeleteWithDeletionVectorsHelper.scala:301]: what is this change for?

Comment (tdas) [reply core/src/main/scala/org/apache/spark/sql/delta/commands/DeleteWithDeletionVectorsHelper.scala:301]: what does this fix? does it have test coverage?

**Review by tdas — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/sql/delta/commands/DeleteWithDeletionVectorsHelper.scala (hunk 15 above)
Comment (allisonport-db) [reply core/src/main/scala/org/apache/spark/sql/delta/commands/DeleteWithDeletionVectorsHelper.scala:301]: https://github.com/delta-io/delta/issues/1725

Tests fail without the fix. Adding a TODO in the code here to provide a more robust solution for 3.4.0 and 3.4.1

**Comment by allisonport-db** [comment]
Closed by 5c3f4d3
